bindings/experimental/distrdf/python/DistRDF/Ranges.py

```diff
@@ -321,82 +324,95 @@ def get_clustered_ranges(clustersinfiles, npartitions, treename, friend_info):
+
+        # The global start and end entries are retrieved as follows:
+        # - Take as start start the `start` attribute of the minimum
+        #   `ChainCluster` object in the current partition
+        # - Take as end the `end` attribute of the maximum `ChainCluster` object
+        #   in the current partition
+        # - Both values need to be shifted by the respective `offset` attribute
+        #   in the same `ChainCluster` instance
+        # - Then to make them "global with respect to the chain of files in the
+        #   current partition", we need to subtract the offset of the first
+        #   cluster in the current partition. That is equal to the amount of
+        #   entries in the chain up to the beginning of that cluster. This is
+        #   needed to maintain a reference of the entries of the range with
+        #   respect to the list of files that hold them. For example, given the
+        #   following files:
+        #         tree10000entries10clusters.root --> 10000 entries, 10 clusters
+        #         tree20000entries10clusters.root --> 20000 entries, 10 clusters
+        #         tree30000entries10clusters.root --> 30000 entries, 10 clusters
+        #   Building 2 ranges will lead to the following tuples::
+        #         TreeRange(globalstart=0,
+        #                 globalend=20000,
+        #                 localstarts = [0, 0],
+        #                 localends = [10000, 10000],
+        #                 filelist=['tree10000entries10clusters.root',
+        #                         'tree20000entries10clusters.root'],
+        #                 friend_info=None)
+        #         TreeRange(globalstart=10000,
+        #                 globalend=50000,
+        #                 localstarts = [10000, 0],
+        #                 localends = [20000, 30000],
+        #                 filelist=['tree20000entries10clusters.root',
+        #                         'tree30000entries10clusters.root'],
+        #                 friend_info=None)
+        #   The first `TreeRange` will read the first 10000 entries from the
+        #   first file, then switch to the second file and read the first 10000
+        #   entries. The second `TreeRange` will start from entry number 10000
+        #   of the second file up until the end of that file (entry number
+        #   20000), then switch to the third file and read the whole 30000
+        #   entries there.
+        firstclusterinpartition = min(partition)
+        partitionoffset = firstclusterinpartition.offset
+
+        lastclusterinpartition = max(partition)
+
+        globalstart = firstclusterinpartition.start # + firstclusterinpartition.offset - partitionoffset
+        globalend = lastclusterinpartition.end + lastclusterinpartition.offset - partitionoffset
+
+        clustered_ranges.append(TreeRange(rangeid, globalstart, globalend, localstarts, localends, filelist, friend_info))
```

Comment:
So here we need to store twice the same kind of info because:
- `globalstart` and `globalend` are needed for `SetCacheEntryRange`
- `localstarts` and `localends` are needed for the construction of the `TEntryLists`

To simplify a bit, alternatively, we could pass `globalstart` and `globalend` plus a list (which maps one to one to `filelist`) that specifies how many entries are in each file. With that information you should be able to construct the TEntryLists.

Reply:
I believe this involves the following changes 
In `Ranges.py`:
```diff
-       localstarts = []
-       localends = []
        filelist = []
+       entriesperfile = []
        for _, clustersinsamefileiter in itertools.groupby(partition, lambda cluster: cluster.filetuple.fileindex):
            clustersinsamefilelist = list(clustersinsamefileiter)

-           localstarts.append(min(clustersinsamefilelist).start)
-           localends.append(max(clustersinsamefilelist).end)
+           entriesperfile.append(max(clustersinsamefilelist).end - min(clustersinsamefilelist).start)
            filelist.append(clustersinsamefilelist[0].filetuple.filename)
```
In `Base.py` (to reconstruct the correct entries to pass to TEntryList
```diff
            if treename is not None:
+                localstarts = []
+                localends = []

+                localstarts.append(treerange.globalstart)
+                localends.append(treerange.globalstart + treerange.entriesperfile.pop(0))

+                for x in treerange.entriesperfile:
+                    localstarts.append(0)
+                    localends.append(x)
```

So overall doesn't look a simplification. Unless I'm overthinking it and there's an even easier way

Reply:
Ok yes, it's a bit more calculation in the worker side for the sake of passing less parameters (with somehow repeated info). Not super important, though. When RDF sequential also does the SetCacheEntryRange itself, we won't need globalstart and globalend anyway, so only the locals will stay.